File: most-active/active.py
"""Find window of time when most authors were active.

For example::

    >>> data = [
    ...    ('Alice', 1901, 1950),
    ...    ('Bob',   1920, 1960),
    ...    ('Carol', 1908, 1945),
    ...    ('Dave',  1951, 1960),
    ... ]

    >>> most_active(data)
    (1920, 1945)

(Alice, Bob, and Carol were all active then).

If there's more than one period, find the earliest::

    >>> data = [
    ...    ('Alice', 1901, 1950),
    ...    ('Bob',   1920, 1960),
    ...    ('Carol', 1908, 1945),
    ...    ('Dave',  1951, 1960),
    ...    ('Eve',   1955, 1985),
    ... ]

    >>> most_active(data)
    (1920, 1945)

(Alice, Bob, Carol were active 1920-1945. Bob, Dave, and Eve were active 1951-1960.
Since there's a tie, the first was returned)
"""


def most_active(bio_data):
    """Find window of time when most authors were active."""

    max_active = 0
    start_year = 0
    end_year = 0
    count_active = False

    for year in range(1900, 2000):
        
        active_authors = 0

        for author in bio_data:
            if author[1] <= year and author[2] >= year:
                active_authors += 1

        if active_authors > max_active:
            max_active = active_authors
            start_year = year
            count_active = True

        if active_authors < max_active and count_active == True:
            end_year = year - 1
            count_active = False

    return (start_year, end_year)


if __name__ == '__main__':
    import doctest
    if doctest.testmod().failed == 0:
        print("\n*** ALL TESTS PASSED. YAY!\n")

File: most-active/active.js
// Find window of time when most authors were active.

// For example::

//     >>> data = [
//     ...    ('Alice', 1901, 1950),
//     ...    ('Bob',   1920, 1960),
//     ...    ('Carol', 1908, 1945),
//     ...    ('Dave',  1951, 1960),
//     ... ]

//     >>> most_active(data)
//     (1920, 1945)

// (Alice, Bob, and Carol were all active then).

// If there's more than one period, find the earliest::

//     >>> data = [
//     ...    ('Alice', 1901, 1950),
//     ...    ('Bob',   1920, 1960),
//     ...    ('Carol', 1908, 1945),
//     ...    ('Dave',  1951, 1960),
//     ...    ('Eve',   1955, 1985),
//     ... ]

//     >>> most_active(data)
//     (1920, 1945)

// (Alice, Bob, Carol were active 1920-1945. Bob, Dave, and Eve were active 1951-1960.
// Since there's a tie, the first was returned)

function mostActive(bioData){
    let maxActive = 0;
    let startYear, endYear, countActive;

    for ( i = 1900; i < 2000; i++){
        
        let numActive = 0;
        
        for ( j = 0; j < bioData.length; j++){
            if (bioData[j][1] <= i && bioData[j][2] >= i) {
                numActive += 1;
            }
        }

        if (numActive > maxActive){
            maxActive = numActive;
            startYear = i;
            countActive = true;
        }

        if (numActive < maxActive && countActive === true){
            countActive = false;
            endYear = i - 1;
        }
    }
    return startYear.toString() + ', ' + endYear.toString();
}